### ISSUE TITLE
Slow `jj status` with large numbers of ignored files

### ISSUE THREAD

**Issue by arxanas** [martinvonz/jj#247]
## Description

`jj status` is taking a while. I'm running it in https://github.com/arxanas/git-branchless, which is not a large repository:

```
$ git ls-files | wc -l
     116

$ git rev-list HEAD --count
816
```

Is there a way to profile this?

## Steps to Reproduce the Problem

1. `jj status`
2. It takes too long.

## Expected Behavior

`jj status` completes in <100ms after first invocation/warming up any caches.

## Actual Behavior

```
$ /usr/bin/time git status
On branch git
nothing to commit, working tree clean
        0.01 real         0.00 user         0.00 sys

$ /usr/bin/time jj status 
Parent commit: 98d9e3d77b9d cleanup: remove deprecated `git.io` link
Working copy : 6473a320cc51 
These branches have conflicts:
  merge
  move-individual-commits
  Use `jj branches` to see details. Use `jj branch <name> -r <rev>` to resolve.
The working copy is clean
        2.71 real         1.16 user         1.53 sys
```

I know that ahead/behind calculations can be slow; not sure if that's a factor here in detecting conflicted branches.

## Specifications

- Platform: macOS 11.6.5
- Version: jj @ 4a1576c87491ba732220dd783fb01c2cf075e8d4

**Comment by arxanas** [comment]
I tried again and got this persistently:

```
$ cargo run --release -- -R ~/Workspace/git-branchless status
    Finished release [optimized] target(s) in 0.21s
     Running `target/release/jj -R /Users/wkhan/workspace/git-branchless status`
thread 'main' panicked at 'failed to create lock file /Users/Waleed/Workspace/git-branchless/.jj/working_copy/working_copy.lock: File exists (os error 17)', lib/src/lock.rs:56:25
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

So I deleted it. I'm not sure if `jj` was running in the background or if it's a stale lock.

This is the release build:

```
$ /usr/bin/time cargo run --release -- -R ~/workspace/git-branchless status
    Finished release [optimized] target(s) in 0.19s
     Running `target/release/jj -R /Users/wkhan/workspace/git-branchless status`
Parent commit: 98d9e3d77b9d cleanup: remove deprecated `git.io` link
Working copy : 6473a320cc51 
These branches have conflicts:
  merge
  move-individual-commits
  Use `jj branches` to see details. Use `jj branch <name> -r <rev>` to resolve.
The working copy is clean
        3.10 real         1.28 user         1.66 sys
```

And here's the debug build 😬:

```
$ /usr/bin/time cargo run -- -R ~/workspace/git-branchless status
    Finished dev [unoptimized + debuginfo] target(s) in 0.18s
     Running `target/debug/jj -R /Users/wkhan/workspace/git-branchless status`
Parent commit: 98d9e3d77b9d cleanup: remove deprecated `git.io` link
Working copy : 6473a320cc51 
These branches have conflicts:
  merge
  move-individual-commits
  Use `jj branches` to see details. Use `jj branch <name> -r <rev>` to resolve.
The working copy is clean
       13.35 real        10.79 user         2.06 sys
```

**Comment by arxanas** [comment]
It doesn't seem like that many:

```
$ git for-each-ref | wc -l
      66
```

(I actually ran a `git branchless gc` as part of this to try to get it to run faster, which cleaned up ~300 refs.)

That being said, I imagine that some of those refs point to commits which are old, so maybe merge-base calculation is involved and slow?

**Comment by martinvonz** [comment]
Can you see how long it takes when not sharing the working copy with git? Something like this:

```
cd $(mktemp -d)
jj init --git-repo=~/Workspace/git-branchless
time jj st
```

**Comment by martinvonz** [comment]
You can also try `jj --no-commit-working-copy st` and see how long that takes. If that's slow, it would indicate that walking the working copy takes a long time. I don't know why it would, though.

**Comment by arxanas** [comment]
There we go:

```
$ /usr/bin/time jj status       
Parent commit: 98d9e3d77b9d cleanup: remove deprecated `git.io` link
Working copy : 6473a320cc51 
These branches have conflicts:
  merge
  move-individual-commits
  Use `jj branches` to see details. Use `jj branch <name> -r <rev>` to resolve.
The working copy is clean
        2.34 real         1.02 user         1.29 sys

$ /usr/bin/time jj --no-commit-working-copy status
Parent commit: 98d9e3d77b9d cleanup: remove deprecated `git.io` link
Working copy : 6473a320cc51 
These branches have conflicts:
  merge
  move-individual-commits
  Use `jj branches` to see details. Use `jj branch <name> -r <rev>` to resolve.
The working copy is clean
        0.01 real         0.00 user         0.00 sys
```

**Comment by martinvonz** [comment]
I think I'm able to reproduce the slowness in the git-branchless repo if I add lots of files in the `target/` directory, so I should be able to figure out what the issue is now.

**Comment by martinvonz** [comment]
It looks like I broke the optimization from c1060610bda276dce60cff69e37f059d80b7d99a in the fix in a028f33e3b21401f894ec996b6ba45998a92df23.
